### PR TITLE
[RPA-1584] Process/Bot notifications scheduler migration

### PR DESCRIPTION
**Pull request by Aeri4a**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

**Review by bhszewczyk — APPROVED** [review]

File: runbotics/runbotics-scheduler/src/database/bot/bot.service.ts [hunk 1]
```diff
@@ -11,7 +11,8 @@ import {
     IBotSystem,
 } from 'runbotics-common';
 
-const relations = ['user', 'system', 'collection', 'subscribers', 'collection.users'];
+// const relations = ['user', 'system', 'collection', 'subscribers', 'collection.users'];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/database/bot/bot.service.ts]: left out comment

File: runbotics/runbotics-scheduler/src/database/process/process.entity.ts [hunk 2]
```diff
@@ -30,6 +30,9 @@ export class ProcessEntity implements IProcess {
     @PrimaryColumn({ type: 'bigint', transformer: numberTransformer })
     id: number;
 
+    @Column({ name: 'tenant_id' })
+    tenantId: string;
+
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/database/process/process.entity.ts:37]: any relation to tenant?

File: runbotics/runbotics-scheduler/src/database/process/process.service.ts [hunk 3]
```diff
@@ -4,7 +4,8 @@ import { Repository } from 'typeorm';
 import { ProcessEntity } from './process.entity';
 import { IProcess } from 'runbotics-common';
 
-const relations = ['createdBy', 'system', 'botCollection', 'schedules', 'editor', 'subscribers.authorities'];
+// const relations = ['createdBy', 'system', 'botCollection', 'schedules', 'editor', 'subscribers.authorities'];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/database/process/process.service.ts]: another comment to remove

File: runbotics/runbotics-scheduler/src/mail/mail.service.ts [hunk 4]
```diff
@@ -70,7 +70,8 @@ export class MailService {
     public async sendBotDisconnectionNotificationMail(bot: IBot, installationId: string) {
         const disconnectedBot = await this.botService.findById(bot.id);
         const botAssignedUserEmail = disconnectedBot.user.email;
-        const subscribers = disconnectedBot.subscribers ?? [];
+        // const subscribers = disconnectedBot.subscribers ?? [];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/mail/mail.service.ts]: redundant comment

File: runbotics/runbotics-scheduler/src/mail/mail.service.ts [hunk 5]
```diff
@@ -96,7 +97,8 @@ export class MailService {
     public async sendProcessFailureNotificationMail(process: IProcess, processInstance: IProcessInstance) {
         const failedProcess = await this.processService.findById(process.id);
         const processCreatorEmail = failedProcess.createdBy.email;
-        const subscribers = failedProcess.subscribers ?? [];
+        // const subscribers = failedProcess.subscribers ?? [];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/mail/mail.service.ts]: redundant

File: runbotics/runbotics-scheduler/src/migrations/1722934541675-GlobalVariablesProcess.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+export class GlobalVariablesProcess1722934541675 implements MigrationInterface {
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "process_global_variable" DROP CONSTRAINT "fk_process_global_variable_process_id"`);
+        await queryRunner.query(`ALTER TABLE "process_global_variable" DROP CONSTRAINT "fk_process_global_variable_global_variable_id"`);
+        await queryRunner.query(`CREATE INDEX "IDX_a5c44b5b12113dbd973043fb6b" ON "process_global_variable" ("process_id") `);
+        await queryRunner.query(`CREATE INDEX "IDX_b6660b0a9bcd79f1fe11e34ac4" ON "process_global_variable" ("global_variable_id") `);
+        await queryRunner.query(`ALTER TABLE "process_global_variable" ADD CONSTRAINT "FK_a5c44b5b12113dbd973043fb6ba" FOREIGN KEY ("process_id") REFERENCES "process"("id") ON DELETE CASCADE ON UPDATE CASCADE`);
+        await queryRunner.query(`ALTER TABLE "process_global_variable" ADD CONSTRAINT "FK_b6660b0a9bcd79f1fe11e34ac4a" FOREIGN KEY ("global_variable_id") REFERENCES "global_variable"("id") ON DELETE CASCADE ON UPDATE CASCADE`);
+    }
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/migrations/1722934541675-GlobalVariablesProcess.ts:12]: everything looks correct

File: runbotics/runbotics-scheduler/src/migrations/1722934554465-NotificationColumnsRescue.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        await queryRunner.query(`ALTER TABLE "notification_bot" ADD CONSTRAINT "fk_user_id_notification_bot" FOREIGN KEY ("user_id") REFERENCES "jhi_user"("id") ON DELETE CASCADE ON UPDATE NO ACTION`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" ADD CONSTRAINT "fk_type_notification_bot" FOREIGN KEY ("type") REFERENCES "notification_type"("type") ON DELETE NO ACTION ON UPDATE NO ACTION`);
+
+        await queryRunner.query(`ALTER TABLE "notification_process" ADD CONSTRAINT "fk_process_id_notification_process" FOREIGN KEY ("process_id") REFERENCES "process"("id") ON DELETE CASCADE ON UPDATE NO ACTION`);
+        await queryRunner.query(`ALTER TABLE "notification_process" ADD CONSTRAINT "fk_user_id_notification_process" FOREIGN KEY ("user_id") REFERENCES "jhi_user"("id") ON DELETE CASCADE ON UPDATE NO ACTION`);
+        await queryRunner.query(`ALTER TABLE "notification_process" ADD CONSTRAINT "fk_type_notification_process" FOREIGN KEY ("type") REFERENCES "notification_type"("type") ON DELETE NO ACTION ON UPDATE NO ACTION`);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/migrations/1722934554465-NotificationColumnsRescue.ts:29]: why are constraints names generated by nest (e.g. FK_63e8bb25d071a7bbd6b80b2503a) dropped but those probably defined in java added (e.g. fk_user_id_notification_process)? It usually works exactly the opposite

File: runbotics/runbotics-scheduler/src/migrations/1722934562965-Notifications.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        await queryRunner.query(`ALTER TABLE "notification_process" DROP CONSTRAINT "fk_user_id_notification_process"`);
+        await queryRunner.query(`ALTER TABLE "notification_process" DROP CONSTRAINT "fk_process_id_notification_process"`);
+        await queryRunner.query(`ALTER TABLE "notification_process" DROP CONSTRAINT "fk_type_notification_process"`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" DROP CONSTRAINT "fk_user_id_notification_bot"`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" DROP CONSTRAINT "fk_bot_id_notification_bot"`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" DROP CONSTRAINT "fk_type_notification_bot"`);
+        await queryRunner.query(`DROP INDEX "public"."IDX_e1a2280a824e267cdb588c74bd"`);
+        await queryRunner.query(`DROP INDEX "public"."IDX_7b592798f4addb0d321581b70b"`);
+        await queryRunner.query(`DROP INDEX "public"."IDX_63e8bb25d071a7bbd6b80b2503"`);
+        await queryRunner.query(`DROP INDEX "public"."IDX_7b63cf21b9a8d158b61d4358ab"`);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/migrations/1722934562965-Notifications.ts:15]: what these indexes are? were're losing them 💔

File: runbotics/runbotics-scheduler/src/migrations/1722934562965-Notifications.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+export class Notifications1722934562965 implements MigrationInterface {
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "notification_process" DROP CONSTRAINT "fk_user_id_notification_process"`);
+        await queryRunner.query(`ALTER TABLE "notification_process" DROP CONSTRAINT "fk_process_id_notification_process"`);
+        await queryRunner.query(`ALTER TABLE "notification_process" DROP CONSTRAINT "fk_type_notification_process"`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" DROP CONSTRAINT "fk_user_id_notification_bot"`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" DROP CONSTRAINT "fk_bot_id_notification_bot"`);
+        await queryRunner.query(`ALTER TABLE "notification_bot" DROP CONSTRAINT "fk_type_notification_bot"`);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/migrations/1722934562965-Notifications.ts:11]: here the order is correct - java constraint names are being dropped (e.g. fk_user_id_notification_process) and nest names (UQ_1c614e65e2b3d2887977548f469) are being added. Maybe 1722934554465-NotificationColumnsRescue.ts and 1722934562965-Notifications.ts should have been merged? (both reverted, both removed and generated new migration, one for both of them)

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        @Body(new ZodValidationPipe(createNotificationBotSchema))
+        notificationBotDto: CreateNotificationBotDto,
+        @User() user: UserEntity
+    ) {
+        this.logger.log(`REST request to create bot notification by user id: ${user.id}`);
+        return this.notificationBotService.create(notificationBotDto, user);
+    }
+
+    @Delete(':id')
+    @HttpCode(HttpStatus.NO_CONTENT)
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts:42]: what's the reason for returning that? Except for reflecting java maybe?

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-process/notification-process.service.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            user: {
+                id: notificationProcess.user.id,
+                login: notificationProcess.user.login
+            },
+            createdAt: notificationProcess.createdAt
+        };
+    }
+
+    private async checkProcessAccessAndGetById(processId: number, user: UserEntity) {
+        // TODO: to be updated after process migration to scheduler
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/notification-process/notification-process.service.ts:76]: You can add info about HOW this should be updated (short, even one sentence, but is it about changing entity import or something else?

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/bot/BotDetailsView/BotConfigure.tsx [hunk 12]
```diff
@@ -41,28 +38,27 @@ const BotConfigure: FC = () => {
         })), [botSubscriptions]);
 
     const handleGetBotSubscribers = async () => {
-        await dispatch(botActions.getBotSubscriptionInfo(botId));
-        await dispatch(botActions.getBotSubscriptionInfoByBotIdAndUserId({ botId, userId }));
+        await dispatch(botActions.getBotSubscriptionInfo({ resourceId: `bots/${botId}` }));
```
Comment (marcin-depcik) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/bot/BotDetailsView/BotConfigure.tsx]: imo `bots/${botId}` doesn't sound like `resourceId`, I would concat `bots/` part with `BOT_NOTIFICATION_PATH` and pass `botId` as the `resourceId`

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessConfigureView/ProcessConfigureView.tsx [hunk 13]
```diff
@@ -65,8 +63,7 @@ const ProcessConfigureView: VFC = () => {
         })), [processSubscriptions]);
 
     const handleGetProcessSubscribers = async () => {
-        await dispatch(processActions.getProcessSubscriptionInfo(processId));
-        await dispatch(processActions.getProcessSubscriptionInfoByProcessIdAndUserId({ processId, userId }));
+        await dispatch(processActions.getProcessSubscriptionInfo({ resourceId: `processes/${processId}` }));
```
Comment (marcin-depcik) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessConfigureView/ProcessConfigureView.tsx]: same here, imo `processes/${processId}` doesn't sound like `resourceId`

File: runbotics/runbotics-scheduler/src/mail/mail.service.ts [hunk 14]
```diff
@@ -70,7 +70,8 @@ export class MailService {
     public async sendBotDisconnectionNotificationMail(bot: IBot, installationId: string) {
         const disconnectedBot = await this.botService.findById(bot.id);
         const botAssignedUserEmail = disconnectedBot.user.email;
-        const subscribers = disconnectedBot.subscribers ?? [];
+        // const subscribers = disconnectedBot.subscribers ?? [];
+        const subscribers = [];
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/mail/mail.service.ts]: why?

File: runbotics/runbotics-scheduler/src/mail/mail.service.ts [hunk 15]
```diff
@@ -96,7 +97,8 @@ export class MailService {
     public async sendProcessFailureNotificationMail(process: IProcess, processInstance: IProcessInstance) {
         const failedProcess = await this.processService.findById(process.id);
         const processCreatorEmail = failedProcess.createdBy.email;
-        const subscribers = failedProcess.subscribers ?? [];
+        // const subscribers = failedProcess.subscribers ?? [];
+        const subscribers = [];
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/mail/mail.service.ts]: why++

File: runbotics/runbotics-scheduler/src/database/bot/bot.entity.ts [hunk 16]
```diff
@@ -52,12 +50,4 @@ export class BotEntity implements IBot {
...
     @JoinColumn([{ name: 'collection_id', referencedColumnName: 'id' }])
     collection: IBotCollection;
-
-    @ManyToMany(() => UserEntity)
-    @JoinTable({
-        name: 'notification_bot',
-        joinColumn: { name: 'bot_id', referencedColumnName: 'id' },
-        inverseJoinColumn: { name: 'user_id', referencedColumnName: 'id' },
-    })
-    subscribers: IUser[];
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/database/bot/bot.entity.ts:62]: why is this not needed?

File: runbotics/runbotics-scheduler/src/database/process/process.entity.ts [hunk 17]
```diff
@@ -82,14 +85,6 @@ export class ProcessEntity implements IProcess {
     @JoinColumn({ name: 'editor_id', referencedColumnName: 'id' })
     editor: IUser;
 
-    @ManyToMany(() => UserEntity)
-    @JoinTable({
-        name: 'notification_process',
-        joinColumn: { name: 'process_id', referencedColumnName: 'id' },
-        inverseJoinColumn: { name: 'user_id', referencedColumnName: 'id' },
-    })
-    subscribers: IUser[];
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/database/process/process.entity.ts:91]: same here

File: runbotics/runbotics-scheduler/src/database/bot/bot.service.ts [hunk 18]
```diff
@@ -11,7 +11,8 @@ import {
...
+const relations = ['user', 'system', 'collection', 'collection.users'];
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/database/bot/bot.service.ts]: why did you remove `subscribers`?

File: runbotics/runbotics-scheduler/src/database/process/process.service.ts [hunk 19]
```diff
@@ -4,7 +4,8 @@ import { Repository } from 'typeorm';
...
+const relations = ['createdBy', 'system', 'botCollection', 'schedules', 'editor'];
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/database/process/process.service.ts]: why did you remove `subscribers.authorities`

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    async deleteNotification(
+        @Param('id', ParseUUIDPipe) id: string,
+        @User() user: UserEntity,
+    ) {
+        this.logger.log('REST request to delete bot notification by id: ', id);
+        await this.notificationBotService.delete(id, user);
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts:48]: you don't need async/await here

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.entity.ts
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.entity.ts:1]: ❤️

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.service.ts [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+                .andWhere('collection.tenantId = :tenantId', { tenantId: user.tenantId })
+                .andWhere(new Brackets(qb => {
+                    qb.where('bot.user.id = :userId')
+                    .orWhere('collection.publicBotsIncluded = true')
+                    .orWhere('collection.createdByUser.id = :userId')
+                    .orWhere('user.id = :userId');
+                })).getOneOrFail().catch(() => {
+                    throw new NotFoundException();
+                });
+    }
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.service.ts:106]: monster query xdd 👹

**Review by marcin-depcik — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/bot/BotDetailsView/BotConfigure.tsx (hunk 12 above)
Comment (Aeri4a) [reply runbotics/runbotics-orchestrator-ui/src/src-app/views/bot/BotDetailsView/BotConfigure.tsx]: I was considering additional property for that kind of situations with ApiTenant wrapper, but for now I think I will add the subresource to path in thunks.

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts (hunk 10 above)
Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts:42]: J A V A :)

File: runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts (hunk 20 above)
Comment (marcin-depcik) [reply runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts:48]: nvm, I was wrong

Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/scheduler-database/notification-bot/notification-bot.controller.ts:48]: Are you sure ? If we return promise then NestJS will resolve it before returning http response. But in this case how does it know if we didn't return after delete method? There could be an error but Nest will return 204... I've tested it and it would be a critical app crash.

**Review by marcin-depcik — APPROVED** [review]
